content/en/blog/_posts/2018-07-09-kubevirt-crds-for-virtualization.md

```diff
@@ -0,0 +1,106 @@
+
+For example, the endpoint to create a VMI object is as follows.
+
+**/apis/kubevirt.io/v1alpha2/namespaces/my-namespace/virtualmachineinstances/my-vm**
+
+However, the subresource endpoint to access graphical VNC looks like this.
+
+**/apis/subresources.kubevirt.io/v1alpha2/namespaces/my-namespace/virtualmachineinstances/my-vm/vnc**
+
+Notice that the first request uses **kubevirt.io** and the second request uses **subresource.kubevirt.io**. We don’t like that, but that’s how we’ve managed to combine CRDs with a stateless aggregated API server for subresources.
+
+One thing worth noting is that in Kubernetes 1.10 a very basic form of CRD subresource support was added in the form of the /status and /scale subresources. This support does not help us deliver the virtualization features we want subresources for. There have however been discussions about exposing custom CRD subresources as webhooks in a future Kubernetes version. If this functionality lands, we will gladly transition away from our stateless aggregated API server workaround to use a subresource webhook feature.
+
+## CRD Finalizers
+
+A [CRD finalizer](https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#advanced-topics) is a feature that lets us provide a pre-delete hook in order to perform actions before allowing a CRD object to be removed from persistent storage. In KubeVirt, we use finalizers to guarantee a virtual machine has completely terminated before we allow the corresponding VMI object to be removed from etcd. 
+
+## API Versioning for CRDs
+
+The Kubernetes core apis have the ability to support multiple versions for a single object type and perform conversions between those versions. This gives the Kubernetes core apis a path for advancing the v1alpha1 version of an object to a v1beta1 version and so forth.
+
+Unfortunately in Kubernetes 1.10, CRDs do not have support for multiple versions. This means when we want to progress a CRD from kubevirt.io/v1alpha1 to kubevirt.io/v1beta1, the only path currently available to us is to backup our CRD objects, delete the registered CRD from Kubernetes, register a new CRD with the updated version, convert the backed up CRD objects to the new version, and finally post the migrated CRD objects back to the cluster.
+
+That strategy is not a viable option for us.
+
+Luckily, while we don’t currently have a complete solution in place to handle CRDs with multiple versions, there is [work underway to rectify this issue in Kubernetes](https://github.com/kubernetes/features/issues/544). Looking at the Kubernetes 1.11 release notes, the [initial steps](https://github.com/kubernetes/kubernetes/pull/63830) have already been taken to kick off this feature. CRDs with multiple versions will be supported soon and we look forward to taking advantage of that feature as soon as it lands.
```

Comment:
This pretty much says that the CRD versioning in 1.11 isn't complete for Kubevirt's use case.  Can you expand on that?  What work remains to be done?  The feature in 1.11 is marked as a complete implementation.

Reply:
> This pretty much says that the CRD versioning in 1.11 isn't complete for Kubevirt's use case. Can you expand on that? What work remains to be done? The feature in 1.11 is marked as a complete implementation.

There's no conversion between CRD versions. It's also strange that all the versions share the same OpenAPIv3 validation. It's like we've created a way to progress an api from alpha->beta->release only if that api contains no changes.

The value of the current implementation seems limited, but it's definitely a first step in the right direction. 

I'll update the doc to reflect my thoughts here.